### ISSUE TITLE
Make Microsoft.Extensions.ServiceDiscovery.UriEndpoint class Public

### ISSUE THREAD

**Issue by thompson-tomo** [dotnet/aspire#4224]
I am using the endpoint resolver to obtain my endpoints however i do not have the option of obtaining a structured uri for my endpoint due to the UriEnpointClass being internal. As such i am needing to cast the endpoint to a string and then construct a new URI using the string. This double conversion should not be necessary given that the URI Endpoint already contains the uri.

The reason which i am needing it as a Uri is so that i can get the path string.

**Comment by pfeigl** [comment]
I ran into this also today and I'm wondering whether my (our) whole approach is wrong?

I'm having two projects. One hosts a NON http service. I mark the port I'm using here as Endpoint in the AppHost.
The other project wants to consume this service and is marked with the relevant reference to the former resource.

Again, this is not http based and thus the ServiceDiscovery on the http message handler wont work here.
Therefor I fell back to use the normal `ServiceEndpointResolver`.

However calling `GetEndpointsAsync` returns `UriEndPoints`. These are as @thompson-tomo outlined internal and even do not implement the `Serialize` method.

I kind of have the feeling I'm doing something completely wrong. Perhaps someone can shed some light on this area, where non HTTP traffic is in the play.

**Comment by kongxy10** [comment]
Agreed. I'm unable to implement a ```IServiceEndpointProvider``` that can resolve a ```UriEndPoint```.

**Comment by bart-vmware** [comment]
I'm running into this limitation as well. It is problematic because the existing logic type-checks for the internal `UriEndPoint` type [here](https://github.com/dotnet/extensions/blob/09472076b1c0d53822d9ce085d9df408286bb867/src/Libraries/Microsoft.Extensions.ServiceDiscovery/Http/ResolvingHttpDelegatingHandler.cs#L65), throwing otherwise, which makes it impossible to define and provide our own type. This basically makes it impossible to handle URIs when implementing service discovery, without resorting to reflection to create a `UriEndPoint` instance.

Related to that, I find myself reflection-loading [this method](https://github.com/dotnet/extensions/blob/09472076b1c0d53822d9ce085d9df408286bb867/src/Libraries/Microsoft.Extensions.ServiceDiscovery/Configuration/ConfigurationServiceEndpointProvider.cs#L202) and [this method](https://github.com/dotnet/extensions/blob/09472076b1c0d53822d9ce085d9df408286bb867/src/Libraries/Microsoft.Extensions.ServiceDiscovery/ServiceDiscoveryOptions.cs#L33) in various projects that implement service discovery.

To address these, I'd like to propose the following public API additions:

1. Make the type `UriEndPoint` public.
   ```c#
   namespace Microsoft.Extensions.ServiceDiscovery
   {
       public class UriEndPoint
       {
           // Same implementation as internal type, non-sealed, adding a null check in the constructor.
       }
   }
   ```
   Note that a similar public type already exists in a different namespace [here](https://github.com/dotnet/extensions/blob/09472076b1c0d53822d9ce085d9df408286bb867/src/Libraries/Microsoft.Extensions.Http.Resilience/Routing/UriEndpoint.cs#L12), whose `Uri` is nullable. I suppose both types can co-exist.

2. Add a static method to construct a `ServiceEndpoint` instance from a `string`.
   ```c#
   namespace Microsoft.Extensions.ServiceDiscovery
   {
       public class ServiceEndpoint
       {
           public static bool TryParse([NotNullWhen(true)] string? value,
               [NotNullWhen(true)] out ServiceEndpoint? serviceEndpoint)
           {
               // Move the implementation of ConfigurationServiceEndpointProvider.TryParseEndPoint to here,
               // calling ServiceEndpoint.Create on its output if not null.
           }
       }
   }
   ```

3. Replace the internal static method `ServiceDiscoveryOptions.ApplyAllowedSchemes` with:
   ```c#
   namespace Microsoft.Extensions.ServiceDiscovery
   {
       public class ServiceDiscoveryOptions
       {
           public IList<string> GetAllowedSchemes(IList<string> schemes)
       }
   }
   ```
   Based on the single usage [here](https://github.com/dotnet/extensions/blob/09472076b1c0d53822d9ce085d9df408286bb867/src/Libraries/Microsoft.Extensions.ServiceDiscovery/Configuration/ConfigurationServiceEndpointProvider.cs#L44), this method does not need to be static and can use instance properties instead.

Would you be interested in such a PR?
/cc @rzikm @ReubenBond @davidfowl

**Comment by rzikm** [comment]
I'm afraid I can't help here, I only contributed the DNS resolver implementation and don't know what were the design considerations behind the ServiceDiscovery API. I hope that @ReubenBond can assist here.

**Comment by bart-vmware** [comment]
@ReubenBond  @davidfowl
I believe this issue can be closed as completed. `UriEndPoint` has been made public in https://github.com/dotnet/extensions/pull/6946.
